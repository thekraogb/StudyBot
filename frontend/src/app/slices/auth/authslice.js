import { createSlice } from "@reduxjs/toolkit";

const authSlice = createSlice({
  name: "auth",
  initialState: {
    name: localStorage.getItem("name") || null,
    email: localStorage.getItem("email") || null,
    token: localStorage.getItem("accessToken") || null,
  },
  reducers: {
    setAccessToken: (state, action) => {
      const { accessToken } = action.payload;
      state.token = accessToken;

      localStorage.setItem("accessToken", accessToken);
    },
    setCredentials: (state, action) => {
      const { name, email } = action.payload;
      state.name = name;
      state.email = email;

      localStorage.setItem("name", name);
      localStorage.setItem("email", email);
    },
    logOut: (state) => {
      state.email = null;
      state.token = null;
      state.name = null;

      localStorage.removeItem("name");
      localStorage.removeItem("email");
      localStorage.removeItem("accessToken");

      localStorage.removeItem("chats");
      localStorage.removeItem("chatId");
      localStorage.removeItem("messages");
    },
  },
});

export const { setAccessToken, setCredentials, logOut } = authSlice.actions;

export default authSlice.reducer;

export const selectCurrentToken = (state) => state.auth.token;
