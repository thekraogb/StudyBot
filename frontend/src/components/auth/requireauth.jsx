import { useLocation, Navigate, Outlet } from "react-router-dom";
import { useSelector } from "react-redux";
import { selectCurrentToken } from "../../app/slices/auth/authslice";

const RequireAuth = () => {
  const token = useSelector(selectCurrentToken);
  const location = useLocation();

  return token ? (
    <Outlet />
  ) : (
    <Navigate to="/" state={{ from: location }} replace />
  );
};
export default RequireAuth;
