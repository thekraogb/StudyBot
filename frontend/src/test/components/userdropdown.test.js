import React from "react";
import { render, screen, fireEvent, act , waitFor} from "@testing-library/react";
import renderer from "react-test-renderer";
import { logOut } from "../../app/slices/auth/authslice";
import UserDropdown from "../../components/dropdown/userdropdown";
import { useSelector } from "react-redux";
import { store } from "../../app/store";
import { Provider } from "react-redux";
import { useDispatch } from "react-redux";
import { MemoryRouter } from "react-router-dom";

afterEach(() => {
  jest.clearAllMocks();
});

jest.mock("react-redux", () => {
    const actual = jest.requireActual("react-redux");
    const dispatch = jest.fn();
    return {
        ...actual,
      useDispatch: () => dispatch,
      useSelector: jest.fn(),
    };
  });

  beforeEach(() => {
    Object.defineProperty(window, 'location', {
      configurable: true,
      value: { reload: jest.fn() },
    });
  });

describe("UserDropdown", () => {
  beforeEach(() => {
    useSelector.mockImplementation((callback) => {
      return callback({
        auth: {
          name: "Thekra",
          email: "thekra@gmail.com",
        },
      });
    });
  });

  beforeEach(() => {
      render(
        <Provider store={store}>
            <MemoryRouter>
          <UserDropdown />
          </MemoryRouter>
        </Provider>
      );
  });

  test("it renders UserDropdown component", async () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <UserDropdown />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("handleLogoutClick function", () => {
  beforeEach(async () => {
      render(
        <Provider store={store}>
          <MemoryRouter>
          <UserDropdown />
          </MemoryRouter>
        </Provider>
      );
  });

  test("it logs out user", async () => {
    const dispatch = useDispatch();

    const logOutBottun = screen.getByRole("button");

    fireEvent.click(logOutBottun);

    await waitFor(() => { expect(dispatch).toHaveBeenCalledWith(logOut());});
    expect(window.location.reload).toHaveBeenCalled();
  });
});
