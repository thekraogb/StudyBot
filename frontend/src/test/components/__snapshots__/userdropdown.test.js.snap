// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdown it renders UserDropdown component 1`] = `
<div
  className="dropdown"
  onClick={[Function]}
>
  <div
    className="dropdown-menu"
  >
    <div
      className="user-info"
    >
      <strong>
        Thekra
        ,
      </strong>
      <br />
      thekra@gmail.com
    </div>
    <hr
      className="separator"
    />
    <div
      className="logout"
    >
      <button
        className="logout-button"
        onClick={[Function]}
      >
        <img
          alt="Logout"
          src="test-file-stub"
          style={
            {
              "height": 23,
              "width": 23,
            }
          }
        />
        <span>
          Log out
        </span>
      </button>
    </div>
  </div>
</div>
`;
